### PR TITLE
PUBDEV-6611: h2o.scale() only applies scaling to numeric columns

### PR DESCRIPTION
**Pull request by michalkurka**

### CODE REVIEW

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java [hunk 1]
```diff
@@ -35,30 +35,69 @@ public String str() {
...
+  private static double[] calcMeans(Env env, AstRoot meanSpec, Frame fr) {
+    final int ncols = fr.numCols();
     double[] means;
-    if (asts[2] instanceof AstNumList) {
-      means = ((AstNumList) asts[2]).expand();
+    if (meanSpec instanceof AstNumList) {
+      means = ((AstNumList) meanSpec).expand();
       if (means.length != ncols)
-        throw new IllegalArgumentException("Numlist must be the same length as the columns of the Frame");
+        throw new IllegalArgumentException("Numlist must be the same length as the numeric columns of the Frame");
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java]: I'd personally test if the exception is thrown in JUnit.

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java [hunk 2]
```diff
@@ -35,30 +35,69 @@ public String str() {
...
+    private final double[] _mults;
+
+    InPlaceSlateTask(double[] means, double[] mults) {
+      _means = means;
+      _mults = mults;
+    }
 
-    // Peel out the bias/shift/mean
+    @Override
+    public void map(Chunk[] cs) {
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java]: Great !

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java [hunk 3]
```diff
@@ -35,30 +35,69 @@ public String str() {
...
+  private static double[] calcMults(Env env, AstRoot multSpec, Frame fr) {
+    final int ncols = fr.numCols();
     double[] mults;
-    if (asts[3] instanceof AstNumList) {
-      mults = ((AstNumList) asts[3]).expand();
+    if (multSpec instanceof AstNumList) {
+      mults = ((AstNumList) multSpec).expand();
       if (mults.length != ncols)
-        throw new IllegalArgumentException("Numlist must be the same length as the columns of the Frame");
+        throw new IllegalArgumentException("Numlist must be the same length as the numeric columns of the Frame");
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java]: JUnit test it ?

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java [hunk 4]
```diff
@@ -35,30 +35,69 @@ public String str() {
...
+      Vec v = originalFrame.vec(i);
+      if (v.get_type() == Vec.T_NUM) {
+        numericFrame.add(originalFrame.name(i), v);
+      }
+    }
+
+    final double[] means = calcMeans(env, asts[2], numericFrame);
+    final double[] mults = calcMults(env, asts[3], numericFrame);
+    // Update in-place.
+    new InPlaceSlateTask(means, mults).doAll(numericFrame);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java]: This call can probably be completely avoided if `numericFrame.ncols() == 0` ?

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java (hunk 1 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java]: Sure, thanks!

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java (hunk 3 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java]: Yes, will add!

**Pull request merged**